File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeList(self, list1, list2):
        # merge 2 lists and return head
        dummy = ListNode()
        curr = dummy
        while list1 and list2:
            if list1.val <= list2.val:
                curr.next = list1
                list1 = list1.next
            else:
                curr.next = list2
                list2 = list2.next
            curr = curr.next
        curr.next = list1 or list2
        return dummy.next

    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if len(lists) ==0:
            return None
        #merge pairs of lists and create new list of lists to merge
        currLists = lists
        newLists = []
        i = 0
        while len(currLists) > 1:
            newLists = []
            i = 0
            while i + 1 < len(currLists):
                newLists.append(self.mergeList(currLists[i], currLists[i+1]))
                i += 2

            if len(currLists) % 2 == 1:
                # add one last list
                newLists.append(currLists[len(currLists) - 1])

            currLists = newLists

        return currLists[0]
        

File: Data Structures & Algorithms/palindromic-substrings/submission-1.cpp
class Solution {
public:
    int countSubstrings(string s) {
        int n = s.length();
        if(n == 1) {
            return 1;
        }
        vector<vector<bool>> dp(n, vector<bool>(n, 0));
        //init
        for (int i = 0; i < n; i++) {
            dp[i][i] = true;
        }
        int count = n;
        for (int len = 2; len <= n; len++) {
            for (int i = 0; i + len - 1 < n; i++) {
                int j = i + len - 1;
                if (s[i] == s[j]) {
                    if (len == 2) {
                        dp[i][j] = true;
                    }
                    else {
                        dp[i][j] = dp[i+1][j-1];
                    }
                }
                if (dp[i][j]) {
                    count++;
                }
            }    
        }
        return count;
    }
};
